File: sdk/typescript/src/mockClient.ts
import { NotFoundError } from './errors.js';
import type {
  Artifact,
  BindingKind,
  CreateCredentialRequest,
  CreateEnvironmentRequest,
  CreateEnvironmentVersionRequest,
  CreateWebhookRequest,
  Credential,
  Environment,
  EnvironmentVersion,
  GetEnvironmentResult,
  VariableType,
  WebhookDelivery,
  WebhookEvent,
  WebhookSubscription,
  WorkspaceSettings,
} from './types.js';
import { randomUUID, createHash } from 'crypto';

/**
 * Mock client that returns fixture data.
 *
 * Use this for development while the backend is being built.
 * All methods return typed fixture data.
 */
export class MockOmoiOSClient {
  private credentials: Map<string, Credential> = new Map();
  private environments: Map<string, Environment> = new Map();
  private environmentVersions: Map<string, EnvironmentVersion[]> = new Map();
  private artifacts: Map<string, Artifact> = new Map();
  private webhooks: Map<string, WebhookSubscription> = new Map();
  private webhookDeliveries: Map<string, WebhookDelivery[]> = new Map();
  private workspaceSettings: Map<string, WorkspaceSettings> = new Map();

  constructor() {
    this.initFixtures();
  }

  private initFixtures(): void {
    const now = new Date().toISOString();

    const cred: Credential = {
      id: 'cred_1',
      workspace_id: 'ws_1',
      kind: 'bearer_secret',
      name: 'test-api-key',
      created_at: now,
      rotated_at: null,
    };
    this.credentials.set(cred.id, cred);

    const env: Environment = {
      id: 'env_1',
      org_id: 'org_1',
      name: 'staging',
      description: 'Staging environment',
      created_at: now,
      updated_at: now,
    };
    this.environments.set(env.id, env);

    const version: EnvironmentVersion = {
      id: 'ver_1',
      environment_id: env.id,
      version_number: 1,
      variables: {
        DB_URL: { type: 'string' as VariableType, value: 'postgres://localhost:5432/staging' },
        API_KEY: { type: 'secret' as VariableType, value: '***' },
      },
      created_at: now,
    };
    this.environmentVersions.set(env.id, [version]);

    const artifact: Artifact = {
      id: 'art_1',
      workspace_id: 'ws_1',
      name: 'test-file.txt',
      storage_backend: 'local',
      storage_path: '/artifacts/art_1',
      checksum: 'sha256:' + 'a'.repeat(64),
      size_bytes: 1024,
      content_type: 'text/plain',
      artifact_metadata: { source: 'test' },
      created_at: now,
      updated_at: now,
    };
    this.artifacts.set(artifact.id, artifact);

    const webhook: WebhookSubscription = {
      id: 'wh_1',
      org_id: 'org_1',
      url: 'https://example.com/webhook',
      events: ['task.completed' as WebhookEvent, 'artifact.uploaded' as WebhookEvent],
      active: true,
      created_at: now,
    };
    this.webhooks.set(webhook.id, webhook);

    const delivery: WebhookDelivery = {
      id: 'wd_1',
      subscription_id: webhook.id,
      event: 'task.completed' as WebhookEvent,
      payload: { task_id: 'task_1', status: 'success' },
      status: 'delivered',
      attempts: 1,
      next_retry_at: null,
      created_at: now,
    };
    this.webhookDeliveries.set(webhook.id, [delivery]);

    const settings: WorkspaceSettings = {
      id: 'ws_settings_1',
      workspace_id: 'ws_1',
      egress_allowlist: ['api.github.com', 'pypi.org'],
      max_artifact_size_mb: 100,
      allowed_binding_kinds: ['bearer_secret' as BindingKind, 'user_oauth' as BindingKind],
    };
    this.workspaceSettings.set(settings.workspace_id, settings);
  }

  private generateId(prefix: string): string {
    return `${prefix}_${randomUUID().replace(/-/g, '').slice(0, 8)}`;
  }

  listCredentials(workspaceId?: string): Credential[] {
    const creds = Array.from(this.credentials.values());
    if (workspaceId) {
      return creds.filter((c) => c.workspace_id === workspaceId);
    }
    return creds;
  }

  getCredential(credentialId: string): Credential {
    const cred = this.credentials.get(credentialId);
    if (!cred) {
      throw new NotFoundError(`Credential not found: ${credentialId}`);
    }
    return cred;
  }

  createCredential(request: CreateCredentialRequest): Credential {
    const now = new Date().toISOString();
    const cred: Credential = {
      id: this.generateId('cred'),
      workspace_id: request.workspace_id ?? 'ws_1',
      kind: request.kind,
      name: request.name,
      created_at: now,
      rotated_at: null,
    };
    this.credentials.set(cred.id, cred);
    return cred;
  }

  deleteCredential(credentialId: string): void {
    if (!this.credentials.has(credentialId)) {
      throw new NotFoundError(`Credential not found: ${credentialId}`);
    }
    this.credentials.delete(credentialId);
  }

  listEnvironments(): Environment[] {
    return Array.from(this.environments.values());
  }

  getEnvironment(environmentId: string): GetEnvironmentResult {
    const env = this.environments.get(environmentId);
    if (!env) {
      throw new NotFoundError(`Environment not found: ${environmentId}`);
    }
    const versions = this.environmentVersions.get(environmentId) ?? [];
    const latest = versions[versions.length - 1] ?? null;
    return { environment: env, latestVersion: latest };
  }

  createEnvironment(request: CreateEnvironmentRequest): Environment {
    const now = new Date().toISOString();
    const env: Environment = {
      id: this.generateId('env'),
      org_id: 'org_1',
      name: request.name,
      description: request.description ?? null,
      created_at: now,
      updated_at: now,
    };
    this.environments.set(env.id, env);
    this.environmentVersions.set(env.id, []);
    return env;
  }

  createEnvironmentVersion(
    environmentId: string,
    request: CreateEnvironmentVersionRequest
  ): EnvironmentVersion {
    if (!this.environments.has(environmentId)) {
      throw new NotFoundError(`Environment not found: ${environmentId}`);
    }

    const versions = this.environmentVersions.get(environmentId) ?? [];
    const versionNumber = versions.length + 1;

    const now = new Date().toISOString();
    const version: EnvironmentVersion = {
      id: this.generateId('ver'),
      environment_id: environmentId,
      version_number: versionNumber,
      variables: request.variables,
      created_at: now,
    };
    versions.push(version);
    this.environmentVersions.set(environmentId, versions);
    return version;
  }

  uploadArtifact(fileContent: Buffer, workspaceId?: string): Artifact {
    const now = new Date().toISOString();
    const artifactId = this.generateId('art');
    const checksum = createHash('sha256').update(fileContent).digest('hex');

    const artifact: Artifact = {
      id: artifactId,
      workspace_id: workspaceId ?? 'ws_1',
      name: 'uploaded-file.bin',
      storage_backend: 'local',
      storage_path: `/artifacts/${artifactId}`,
      checksum: `sha256:${checksum}`,
      size_bytes: fileContent.length,
      content_type: 'application/octet-stream',
      artifact_metadata: null,
      created_at: now,
      updated_at: now,
    };
    this.artifacts.set(artifact.id, artifact);
    return artifact;
  }

  listArtifacts(workspaceId?: string): Artifact[] {
    const artifacts = Array.from(this.artifacts.values());
    if (workspaceId) {
      return artifacts.filter((a) => a.workspace_id === workspaceId);
    }
    return artifacts;
  }

  getArtifact(artifactId: string): Artifact {
    const artifact = this.artifacts.get(artifactId);
    if (!artifact) {
      throw new NotFoundError(`Artifact not found: ${artifactId}`);
    }
    return artifact;
  }

  downloadArtifact(artifactId: string): Buffer {
    if (!this.artifacts.has(artifactId)) {
      throw new NotFoundError(`Artifact not found: ${artifactId}`);
    }
    return Buffer.from('mock artifact content');
  }

  deleteArtifact(artifactId: string): void {
    if (!this.artifacts.has(artifactId)) {
      throw new NotFoundError(`Artifact not found: ${artifactId}`);
    }
    this.artifacts.delete(artifactId);
  }

  listWebhooks(): WebhookSubscription[] {
    return Array.from(this.webhooks.values());
  }

  createWebhook(request: CreateWebhookRequest): WebhookSubscription {
    const now = new Date().toISOString();
    const webhook: WebhookSubscription = {
      id: this.generateId('wh'),
      org_id: 'org_1',
      url: request.url,
      events: request.events,
      active: true,
      created_at: now,
    };
    this.webhooks.set(webhook.id, webhook);
    this.webhookDeliveries.set(webhook.id, []);
    return webhook;
  }

  deleteWebhook(webhookId: string): void {
    if (!this.webhooks.has(webhookId)) {
      throw new NotFoundError(`Webhook not found: ${webhookId}`);
    }
    this.webhooks.delete(webhookId);
    this.webhookDeliveries.delete(webhookId);
  }

  listWebhookDeliveries(subscriptionId: string): WebhookDelivery[] {
    if (!this.webhooks.has(subscriptionId)) {
      throw new NotFoundError(`Webhook subscription not found: ${subscriptionId}`);
    }
    return this.webhookDeliveries.get(subscriptionId) ?? [];
  }

  getWorkspaceSettings(workspaceId: string): WorkspaceSettings {
    const settings = this.workspaceSettings.get(workspaceId);
    if (!settings) {
      throw new NotFoundError(`Workspace not found: ${workspaceId}`);
    }
    return settings;
  }
}

File: sdk/python/omoios/mock_client.py
"""Mock client for OmoiOS API - returns fixture data for development."""

import hashlib
import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional, Union

from omoios.client import OmoiOSClient
from omoios.exceptions import NotFoundError
from omoios.types import (
    Artifact,
    BindingKind,
    CreateCredentialRequest,
    CreateEnvironmentRequest,
    CreateEnvironmentVersionRequest,
    CreateWebhookRequest,
    Credential,
    Environment,
    EnvironmentVariable,
    EnvironmentVersion,
    VariableType,
    WebhookDelivery,
    WebhookEvent,
    WebhookSubscription,
    WorkspaceSettings,
)


class MockOmoiOSClient(OmoiOSClient):
    """Mock client that returns fixture data.

    Use this for development while the backend is being built.
    All methods return typed fixture data.
    """

    def __init__(self):
        """Initialize mock client with fixture data."""
        super().__init__(base_url="https://api.omoios.dev", api_key="mock-key")
        self._credentials: Dict[str, Credential] = {}
        self._environments: Dict[str, Environment] = {}
        self._environment_versions: Dict[str, List[EnvironmentVersion]] = {}
        self._artifacts: Dict[str, Artifact] = {}
        self._webhooks: Dict[str, WebhookSubscription] = {}
        self._webhook_deliveries: Dict[str, List[WebhookDelivery]] = {}
        self._workspace_settings: Dict[str, WorkspaceSettings] = {}

        # Initialize with fixture data
        self._init_fixtures()

    def _init_fixtures(self) -> None:
        """Initialize fixture data."""
        now = datetime.now()

        # Credentials
        cred = Credential(
            id="cred_1",
            workspace_id="ws_1",
            kind=BindingKind.BEARER_SECRET,
            name="test-api-key",
            config={},
            version=1,
            created_at=now,
            rotated_at=None,
        )
        self._credentials[cred.id] = cred

        # Environments
        env = Environment(
            id="env_1",
            org_id="org_1",
            name="staging",
            description="Staging environment",
            created_at=now,
            updated_at=now,
        )
        self._environments[env.id] = env

        # Environment version
        version = EnvironmentVersion(
            id="ver_1",
            environment_id=env.id,
            version_number=1,
            variables={
                "DB_URL": EnvironmentVariable(
                    type=VariableType.STRING, value="postgres://localhost:5432/staging"
                ),
                "API_KEY": EnvironmentVariable(type=VariableType.SECRET, value="***"),
            },
            created_at=now,
        )
        self._environment_versions[env.id] = [version]

        # Artifacts
        artifact = Artifact(
            id="art_1",
            workspace_id="ws_1",
            name="test-file.txt",
            storage_backend="local",
            storage_path="/artifacts/art_1",
            checksum="sha256:" + "a" * 64,
            size_bytes=1024,
            content_type="text/plain",
            artifact_metadata={"source": "test"},
            created_at=now,
            updated_at=now,
        )
        self._artifacts[artifact.id] = artifact

        # Webhooks
        webhook = WebhookSubscription(
            id="wh_1",
            org_id="org_1",
            url="https://example.com/webhook",
            events=[WebhookEvent.TASK_COMPLETED, WebhookEvent.ARTIFACT_UPLOADED],
            active=True,
            created_at=now,
        )
        self._webhooks[webhook.id] = webhook

        # Webhook deliveries
        delivery = WebhookDelivery(
            id="wd_1",
            subscription_id=webhook.id,
            event=WebhookEvent.TASK_COMPLETED,
            payload={"task_id": "task_1", "status": "success"},
            status="delivered",
            attempts=1,
            next_retry_at=None,
            created_at=now,
        )
        self._webhook_deliveries[webhook.id] = [delivery]

        # Workspace settings
        settings = WorkspaceSettings(
            id="ws_settings_1",
            workspace_id="ws_1",
            egress_allowlist=["api.github.com", "pypi.org"],
            max_artifact_size_mb=100,
            allowed_binding_kinds=[
                BindingKind.BEARER_SECRET,
                BindingKind.USER_OAUTH,
            ],
        )
        self._workspace_settings[settings.workspace_id] = settings

    # Credentials

    def list_credentials(self, workspace_id: Optional[str] = None) -> List[Credential]:
        """List credentials."""
        creds = list(self._credentials.values())
        if workspace_id:
            creds = [c for c in creds if c.workspace_id == workspace_id]
        return creds

    def get_credential(self, credential_id: str) -> Credential:
        """Get a credential by ID."""
        if credential_id not in self._credentials:
            raise NotFoundError(f"Credential not found: {credential_id}")
        return self._credentials[credential_id]

    def create_credential(self, request: CreateCredentialRequest) -> Credential:
        """Create a new credential."""
        now = datetime.now()
        cred = Credential(
            id=f"cred_{uuid.uuid4().hex[:8]}",
            workspace_id=request.workspace_id,
            kind=request.kind,
            name=request.name,
            config=request.config or {},
            version=1,
            created_at=now,
            rotated_at=None,
        )
        self._credentials[cred.id] = cred
        return cred

    def delete_credential(self, credential_id: str) -> None:
        """Delete a credential."""
        if credential_id not in self._credentials:
            raise NotFoundError(f"Credential not found: {credential_id}")
        del self._credentials[credential_id]

    # Environments

    def list_environments(self) -> List[Environment]:
        """List all environments."""
        return list(self._environments.values())

    def get_environment(
        self, environment_id: str
    ) -> Dict[str, Union[Environment, Optional[EnvironmentVersion]]]:
        """Get environment with latest version."""
        if environment_id not in self._environments:
            raise NotFoundError(f"Environment not found: {environment_id}")
        env = self._environments[environment_id]
        versions = self._environment_versions.get(environment_id, [])
        latest = versions[-1] if versions else None
        return {"environment": env, "latestVersion": latest}

    def create_environment(self, request: CreateEnvironmentRequest) -> Environment:
        """Create a new environment."""
        now = datetime.now()
        env = Environment(
            id=f"env_{uuid.uuid4().hex[:8]}",
            org_id=request.org_id,
            name=request.name,
            description=request.description,
            created_at=now,
            updated_at=now,
        )
        self._environments[env.id] = env
        self._environment_versions[env.id] = []
        return env

    def create_environment_version(
        self, environment_id: str, request: CreateEnvironmentVersionRequest
    ) -> EnvironmentVersion:
        """Create a new environment version."""
        if environment_id not in self._environments:
            raise NotFoundError(f"Environment not found: {environment_id}")

        versions = self._environment_versions.get(environment_id, [])
        version_number = len(versions) + 1

        version = EnvironmentVersion(
            id=f"ver_{uuid.uuid4().hex[:8]}",
            environment_id=environment_id,
            version_number=version_number,
            variables=request.variables,
            created_at=datetime.now(),
        )
        versions.append(version)
        self._environment_versions[environment_id] = versions
        return version

    # Artifacts

    def upload_artifact(
        self, file_content: bytes, workspace_id: Optional[str] = None
    ) -> Artifact:
        """Upload an artifact."""
        now = datetime.now()
        artifact_id = f"art_{uuid.uuid4().hex[:8]}"
        checksum = hashlib.sha256(file_content).hexdigest()

        artifact = Artifact(
            id=artifact_id,
            workspace_id=workspace_id or "ws_1",
            name="uploaded-file.bin",
            storage_backend="local",
            storage_path=f"/artifacts/{artifact_id}",
            checksum=f"sha256:{checksum}",
            size_bytes=len(file_content),
            content_type="application/octet-stream",
            artifact_metadata=None,
            created_at=now,
            updated_at=now,
        )
        self._artifacts[artifact.id] = artifact
        return artifact

    def list_artifacts(self, workspace_id: Optional[str] = None) -> List[Artifact]:
        """List artifacts."""
        artifacts = list(self._artifacts.values())
        if workspace_id:
            artifacts = [a for a in artifacts if a.workspace_id == workspace_id]
        return artifacts

    def get_artifact(self, artifact_id: str) -> Artifact:
        """Get an artifact by ID."""
        if artifact_id not in self._artifacts:
            raise NotFoundError(f"Artifact not found: {artifact_id}")
        return self._artifacts[artifact_id]

    def download_artifact(self, artifact_id: str) -> bytes:
        """Download artifact content."""
        if artifact_id not in self._artifacts:
            raise NotFoundError(f"Artifact not found: {artifact_id}")
        # Return mock content
        return b"mock artifact content"

    def delete_artifact(self, artifact_id: str) -> None:
        """Delete an artifact."""
        if artifact_id not in self._artifacts:
            raise NotFoundError(f"Artifact not found: {artifact_id}")
        del self._artifacts[artifact_id]

    # Webhooks

    def list_webhooks(self) -> List[WebhookSubscription]:
        """List webhook subscriptions."""
        return list(self._webhooks.values())

    def create_webhook(self, request: CreateWebhookRequest) -> WebhookSubscription:
        """Create a webhook subscription."""
        now = datetime.now()
        webhook = WebhookSubscription(
            id=f"wh_{uuid.uuid4().hex[:8]}",
            org_id="org_1",
            url=request.url,
            events=request.events,
            active=True,
            created_at=now,
        )
        self._webhooks[webhook.id] = webhook
        self._webhook_deliveries[webhook.id] = []
        return webhook

    def delete_webhook(self, webhook_id: str) -> None:
        """Delete a webhook subscription."""
        if webhook_id not in self._webhooks:
            raise NotFoundError(f"Webhook not found: {webhook_id}")
        del self._webhooks[webhook_id]
        del self._webhook_deliveries[webhook_id]

    def list_webhook_deliveries(self, subscription_id: str) -> List[WebhookDelivery]:
        """List webhook deliveries for a subscription."""
        if subscription_id not in self._webhooks:
            raise NotFoundError(f"Webhook subscription not found: {subscription_id}")
        return self._webhook_deliveries.get(subscription_id, [])

    # Workspace Settings

    def get_workspace_settings(self, workspace_id: str) -> WorkspaceSettings:
        """Get workspace settings."""
        if workspace_id not in self._workspace_settings:
            raise NotFoundError(f"Workspace not found: {workspace_id}")
        return self._workspace_settings[workspace_id]
